pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package middleware
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/server"
+	"google.golang.org/grpc"
+)
+
+// InitGRPCMiddleware initializes the stuff
+func InitGRPCMiddleware(cfg *server.Config) {
+	cfg.GRPCMiddleware = append(cfg.GRPCMiddleware, OpenTracingHTTPGRPCUnaryServerInterceptor(opentracing.GlobalTracer()))
+}
+
+const httpGRPCHandleMethod = "/httpgrpc.HTTP/Handle"
+
+// OpenTracingHTTPGRPCUnaryServerInterceptor returns a grpc.UnaryServerInterceptor suitable
+// for use in a grpc.NewServer call.
+//
+// For example:
+//
+//	s := grpc.NewServer(
+//	    ...,  // (existing ServerOptions)
+//	    grpc.UnaryInterceptor(otgrpc.OpenTracingServerInterceptor(tracer)))
+//
+// All gRPC server spans will look for an OpenTracing SpanContext in the gRPC
+// metadata; if found, the server span will act as the ChildOf that RPC
+// SpanContext.
+//
+// Root or not, the server Span will be embedded in the context.Context for the
+// application-specific gRPC handler(s) to access.
+func OpenTracingHTTPGRPCUnaryServerInterceptor(tracer opentracing.Tracer) grpc.UnaryServerInterceptor {
```

Comment:
Do we need an interceptor for this? Or is this something that could be done directly in the handler for `httpgrpc` requests? (If so, this will likely require moving these changes to `weaveworks/common/httpgrpc`.)

Reply:
I think setting tracing spans and metadata is much more appropriate for middleware/interceptors, rather than the handlers that are doing application logic.

But there are likely some enhancements that we could add to the library itself to make this integrate a little better

Reply:
Sorry, I should have been clearer - when I said "handler for `httpgrpc` requests", I meant the handler in the `httpgrpc` server (ie. [here](https://github.com/weaveworks/common/blob/dd9e68f319d5ddfb68002a238d481265af23fa3b/httpgrpc/server/server.go#L48)), not the application itself.